### PR TITLE
Implements Base Solution Acceleration Methods (#4717)

### PR DESCRIPTION
**Pull request by bscollin**
Implements a base class for relaxed Picard, Modified Picard, and AndersonAcceleration as described in:

https://code.ornl.gov/vera/vera-dev/-/issues/4717

### CODE REVIEW

**Comment by bscollin** [comment]
Do not merge until MPACT change is ready to merge.

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:12]: Typically we have another line of `!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!` before the actual code starts...

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:56]: You could actually do `PROCEDURE(clear_base_SolutionAccelerationType)` here and not define the abstract interface.  Reduces a few lines of code and basically says "the concrete clear routines need to have the same interface as the base clear routine"

You might be able to do the same thing with `init`, but I haven't gotten far enough down to see if that interface holds or not

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:152]: no `s` at the end

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:161]: should have these in the order they appear in

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:208]: Need to set `isInit` to false

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:228]: I'm ok with this being optional... is `1.0` the right default though?

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:231]: do we want a limit on over-relaxation too?  Or just leave it to the client not to be insane?

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:247]: Oh, I see it now... really these should probably just be in `clear_base` in my opinion... calling `init_base` is not sufficient to set `isInit=.TRUE.`, but calling clear_base is definitely sufficient to set it to `.FALSE.`

**Review by aarograh — CHANGES_REQUESTED** [review]
Overall looks good.  A few comments.  I'm also not quite done looking through it but my computer's about to force a restart and I didn't want the review to be weird from stopping in the the middle

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:56]: But if the base class doesn't have the interface defined, then we'll need select types everywhere to handle.  Or are you saying use clear and clear_base as the same?

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:384]: Is it required that `setInital` get called before `step`?  If not, then I think this call might fail due to `%tmpvec` not yet being allocated

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:161]: I just did it the same direction @jpjones6 did it :)

File: src/SolutionAcceleration.f90
Comment (aarograh) [inline src/SolutionAcceleration.f90:391]: Do we really need `%tmpvec` on the class?  Seems like it could just be a local variable for this routine

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:208]: I thought it should be part of the extending class to set isInit to false.  This is why I didn't duplicate it as the main clear routine.  My view is isInit shouldn't be set to false until everything is cleared.

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:228]: I think so, it defaults to a standard Picard iteration.  0 is not an option.  You could maybe argue 0.5, but I don't have a huge preference.

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:231]: There are definitely many cases where overrelaxation makes sense.  We could probably limit it to something like 10, but at this point of the code, I don't really want to be limiting the use.

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:384]: Yes, setInitial needs to be called first.  Honestly, this is consistent with the behavior of Anderson only you had to call reset first in Anderson.

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:391]: Because it could potentially be a large memory allocation depending on vector size... and potentially a distributed vector, I thought it would be better to store it then determine the polymorphic state of a vector and allocate it and destroy it each time.  It was potentially an optimization without a real need, but it would definitely get rid of some code.

File: src/SolutionAcceleration.f90
Comment (aarograh) [reply src/SolutionAcceleration.f90:56]: I'm saying you can pass an actual subroutine name into the `PROCEDURE(...)`.  You only need the abstract interface `clear_absintfc` if there's no routine in the module that defines the right interface.

Another option is to not have `clear` be deferred... just point it at `clear_base_SolutionAccelerationType` and let it get overridden later.  That's probably simpler actually

File: src/SolutionAcceleration.f90
Comment (aarograh) [reply src/SolutionAcceleration.f90:161]: leave things better than you found them ;)

File: src/SolutionAcceleration.f90
Comment (aarograh) [reply src/SolutionAcceleration.f90:208]: Hmm... my view is that `isInit` should only be true if the object is fully initialized... and therefore can be set to false if any part of the object has been cleared...

File: src/SolutionAcceleration.f90
Comment (aarograh) [reply src/SolutionAcceleration.f90:228]: yeah, ok, I'm good with a non-relaxed Picard iteration being the default.  Just wanted to ask.

File: src/SolutionAcceleration.f90
Comment (aarograh) [reply src/SolutionAcceleration.f90:231]: ok, works for me

File: src/SolutionAcceleration.f90
Comment (aarograh) [reply src/SolutionAcceleration.f90:384]: Ok... if we're going to require `setInitial` to be called, then `step` should really have some sort of DBC statement ensuring that the vector is set up.

That DBC statement should really be in the other 2 classes too, even if it's not strictly required.  We should ensure the same set of calls for all classes.

File: src/SolutionAcceleration.f90
Comment (aarograh) [reply src/SolutionAcceleration.f90:391]: That's fair... I think my bigger concern is just that we seem to have required `setInitial` to be called for this class and not others, and making `tmpVec` completely local would resolve that... I could be wrong on the other classes though

Comment (bscollin) [reply src/SolutionAcceleration.f90:391]: yeah, they all need `x(:)` to be allocated too... which `setInitial` also handles

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:208]: Yeah, that's fair enough.

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:12]: Done

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:152]: Fixed

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:161]: Fixed both here and in Anderson

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:208]: Fixed

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:247]: Fixed

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:391]: we good on this one?

File: src/SolutionAcceleration.f90
Comment (jpjones6) [reply src/SolutionAcceleration.f90:152]: Definitely wouldn't follow my example :P.

File: src/SolutionAcceleration.f90
Comment (jpjones6) [inline src/SolutionAcceleration.f90:174]: Greater than 0.

File: src/SolutionAcceleration.f90
Comment (jpjones6) [inline src/SolutionAcceleration.f90:222]: Leaving open the possibility of alpha > 1.0 are we? How devious.

File: src/SolutionAcceleration.f90
Comment (jpjones6) [inline src/SolutionAcceleration.f90:244]: Seems like you changed to just x, or maybe never changed to x_new and meant to...

File: src/SolutionAcceleration.f90
Comment (jpjones6) [inline src/SolutionAcceleration.f90:345]: x

File: src/SolutionAcceleration.f90
Comment (jpjones6) [reply src/SolutionAcceleration.f90:381]: DBC checks to guard against what Aaron mentioned would be nice. Just need one on `tmpVec` I think.

File: src/AndersonAcceleration.f90
Comment (jpjones6) [inline src/AndersonAcceleration.f90:37]: Seems like the mixing parameter for Anderson (beta) and the under-relaxation parameter (alpha) for the Picard classes could just be put on the base class as beta and combined. They both serve the same function.

Comment (bscollin) [reply src/AndersonAcceleration.f90:37]: my basic thought on this was they are called different things in the literature.  I also don't think there's any literature out there that would allow beta > 1, but there'a a lot out there on overrelaxing.  I'm not just thinking about this for coupling, but potentially replacing our current relaxation for currents in cmfd, etc.  Basically anywhere we have a relaxation applied, to use this class... then allow us to more easily experiment with trying Anderson in other areas.

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:374]: Added

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:381]: added DBC checks

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:174]: Fixed

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:222]: yes

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:244]: copy/paste error.  Fixed

File: src/SolutionAcceleration.f90
Comment (bscollin) [reply src/SolutionAcceleration.f90:345]: Fixed

File: src/AndersonAcceleration.f90
Comment (jpjones6) [reply src/AndersonAcceleration.f90:37]: Okay. I'm good with leaving them differentiated.

File: src/SolutionAcceleration.f90
Comment (jpjones6) [reply src/SolutionAcceleration.f90:386]: Looks good to me. @aarograh are you happy with this one?

File: src/SolutionAcceleration.f90
Comment (jpjones6) [reply src/SolutionAcceleration.f90]: After looking at the next routine I figured as much. All good now.

**Review by aarograh — APPROVED** [review]

**Comment by aarograh** [comment]
This is ready once MPACT is ready.

**Pull request merged**
